### PR TITLE
Merge did:indy support

### PR DESCRIPTION
**Pull request by andrewwhitehead**
Currently, tests are failing with the JS wrapper due to a timeout in GetAttribRequest. I think this might be a compatibility issue with the did:indy-supporting indy-node version, but I'm not certain.

### CODE REVIEW

**Comment by swcurran** [comment]
Who can/should look at this?  @blu3beri — is that something you can look at?

@andrewwhitehead — any details that would help — any details on how to reproduce, what is happening/where?

**Comment by andrewwhitehead** [comment]
There are more details in the CI logs, specifically

```
FAIL tests/GetAttribRequest.test.ts (120.218 s)
  GetAttribRequest
    ✕ Submit request (120001 ms)

  ● GetAttribRequest › Submit request

    thrown: "Exceeded timeout of 120000 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

      10 |   beforeAll(() => (pool = setupPool()))
      11 |
    > 12 |   test('Submit request', async () => {
         |   ^
      13 |     const request = new GetAttribRequest({ targetDid: DID, raw: 'endpoint' })
      14 |     const response: GetAttribResponse = await pool.submitRequest(request)
      15 |

      at tests/GetAttribRequest.test.ts:12:3
      at Object.<anonymous> (tests/GetAttribRequest.test.ts:7:1)
```

But there also seems to be a segfault after the tests complete (could be related to the Python test shutdown issue where it needs to stop the tokio threads first, or could be the additional parameters to the `indy_vdr_build_nym_request` FFI method).

**Comment by c2bo** [comment]
This still has a custom old build of indy-node in it. (from here: https://github.com/domwoe/indy-node/tree/feature/did-indy-new). Should we test #112 on this branch and figure out of this fixes things?

**Comment by c2bo** [comment]
Hmm, I tried testing this locally with the current and updated docker image and it seems to have no impact (both times the GetAttribRequest timeouts). Interesting is that I see another test failing:
```
 FAIL  tests/GetNymRequest.test.ts
  ● GetNymRequest › Submit request

    Invalid pointer for result value

      14 |   const indyVdrErrorObject = JSON.parse(nativeError.deref() as string) as IndyVdrErrorObject
      15 |
    > 16 |   throw new IndyVdrError(indyVdrErrorObject)
         |         ^
      17 | }
      18 |

      at handleError (src/error.ts:16:9)
      at NodeJSIndyVdr.buildGetNymRequest (src/NodeJSIndyVdr.ts:267:5)
      at new buildGetNymRequest (../indy-vdr-shared/src/builder/GetNymRequest.ts:28:28)
      at _callee$ (tests/GetNymRequest.test.ts:13:21)
      at tryCatch (../node_modules/regenerator-runtime/runtime.js:63:40)
      at Generator.invoke [as _invoke] (../node_modules/regenerator-runtime/runtime.js:294:22)
      at Generator.next (../node_modules/regenerator-runtime/runtime.js:119:21)
      at tryCatch (../node_modules/regenerator-runtime/runtime.js:63:40)
      at invoke (../node_modules/regenerator-runtime/runtime.js:155:20)
      at ../node_modules/regenerator-runtime/runtime.js:190:11
```

**Comment by c2bo** [comment]
Nevermind - the typescript tests are running with the sovrin buildernet instead of the local pool setup? 
The libindy_vdr header (in libindy_vdr/include) file seems to be the old one before the changes - shouldn't this always create problems with a changed interface?

**Comment by andrewwhitehead** [comment]
@c2bo Not sure if the separate did-indy-pool.dockerfile is still needed or we can just use indy-node-container 1.13.2-rc5 for all tests?

**Comment by c2bo** [comment]
> @c2bo Not sure if the separate did-indy-pool.dockerfile is still needed or we can just use indy-node-container 1.13.2-rc5 for all tests?

We don't need the separate dockerifle anymore as far as i know -> That would be fixed by #112. I will re-base to this branch and create a new PR

**Comment by TimoGlastra** [comment]
We'll try to update the js wrapper to use the local pool instead!

**Comment by c2bo** [comment]
I added a small bash script in #169 that we can use for the different tests that want to use a local node pool. I just realized the same is true for the python wrapper - that one currently uses the idunion test network for its tests
```
./test.sh up
./test.sh down
```

File: test.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#/bin/bash
```
Comment (berendsliedrecht) [inline test.sh:1]: Do we want to keep this file? If so, might be good to move it inside a scripts directory.

File: libindy_vdr/src/common/error.rs [hunk 2]
```diff
@@ -46,6 +46,8 @@ pub enum VdrErrorKind {
     PoolRequestFailed(String),
     #[error("Pool timeout")]
     PoolTimeout,
+    #[error("Resolver error ")]
```
Comment (berendsliedrecht) [inline libindy_vdr/src/common/error.rs:49]:
```suggestion
    #[error("Resolver error")]
```

File: libindy_vdr/src/ffi/resolver.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#[no_mangle]
+pub extern "C" fn indy_vdr_resolve(
+    pool_handle: PoolHandle,
+    did: FfiStr,
+    cb: Option<extern "C" fn(cb_id: i64, err: ErrorCode, response: *const c_char)>,
+    cb_id: i64,
+) -> ErrorCode {
+    catch_err! {
+    trace!("Resolve DID: {:#?}", did);
+    let cb = cb.ok_or_else(|| input_err("No callback provided")).unwrap();
```
Comment (berendsliedrecht) [inline libindy_vdr/src/ffi/resolver.rs:18]:
```suggestion
    let cb = cb.ok_or_else(|| input_err("No callback provided"))?;
```

File: libindy_vdr/src/ffi/resolver.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                Ok(result) => (ErrorCode::Success, result),
+                Err(err) => {
+
+                    let code = ErrorCode::from(err.kind());
+                    set_last_error(Some(err));
+                    (code, String::new())
+
+                }
+            };
+            cb(cb_id, errcode, rust_string_to_c(reply))
```
Comment (berendsliedrecht) [inline libindy_vdr/src/ffi/resolver.rs:38]:
```suggestion
        Box::new(move |ledger_reply| {
            let (errcode, reply) = match handle_resolution_result(ledger_reply, did) {
                Ok(result) => (ErrorCode::Success, result),
                Err(err) => {
                    let code = ErrorCode::from(err.kind());
                    set_last_error(Some(err));
                    (code, String::new())
                }
            };
            cb(cb_id, errcode, rust_string_to_c(reply))
```

File: libindy_vdr/src/ffi/resolver.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#[no_mangle]
+pub extern "C" fn indy_vdr_dereference(
+    pool_handle: PoolHandle,
+    did_url: FfiStr,
+    cb: Option<extern "C" fn(cb_id: i64, err: ErrorCode, response: *const c_char)>,
+    cb_id: i64,
+) -> ErrorCode {
+    catch_err! {
+    trace!("Dereference DID Url: {:#?}", did_url);
+    let cb = cb.ok_or_else(|| input_err("No callback provided")).unwrap();
```
Comment (berendsliedrecht) [inline libindy_vdr/src/ffi/resolver.rs:54]:
```suggestion
    let cb = cb.ok_or_else(|| input_err("No callback provided"))?;
```

File: libindy_vdr/src/ffi/resolver.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+                    let (errcode, reply) = match handle_resolution_result(ledger_reply, did_url) {
+                        Ok(result) => (ErrorCode::Success, result),
+                        Err(err) => {
+
+                            let code = ErrorCode::from(err.kind());
+                            set_last_error(Some(err));
+                            (code, String::new())
+
+                        }
```
Comment (berendsliedrecht) [inline libindy_vdr/src/ffi/resolver.rs:72]:
```suggestion
                Box::new(move |ledger_reply| {
                    let (errcode, reply) = match handle_resolution_result(ledger_reply, did_url) {
                        Ok(result) => (ErrorCode::Success, result),
                        Err(err) => {
                            let code = ErrorCode::from(err.kind());
                            set_last_error(Some(err));
                            (code, String::new())
                        }
```

File: libindy_vdr/src/resolver/resolver.rs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        } else {
+            None
+        };
+
+        let diddoc = match data {
+            Result::DidDocument(mut doc) => {
+                // Try to find legacy endpoint using a GET_ATTRIB txn if diddoc_content is none
+                if doc.diddoc_content.is_none() {
+                    let (seq_no, timestamp) = if let Some(md) = &md {
+                        let j_seq_no = &md.node_response["result"]["seqNo"];
```
Comment (berendsliedrecht) [inline libindy_vdr/src/resolver/resolver.rs:67]: More out of curiousity, what is the `j_` prefix?

File: libindy_vdr/src/resolver/utils.rs [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+                object_type: String::from("Unknown"),
+            }),
+        ),
+    })
+}
+
+pub fn parse_ledger_data(ledger_data: &str) -> VdrResult<(SJsonValue, SJsonValue, SJsonValue)> {
+    let v: SJsonValue = serde_json::from_str(ledger_data)
+        .map_err(|_| err_msg(VdrErrorKind::Resolver, "Could not parse ledger response"))?;
+    // Unwrap should be safe here
```
Comment (berendsliedrecht) [inline libindy_vdr/src/resolver/utils.rs:230]: Nothing is being unwrapped anymore.

File: libindy_vdr/tests/nym.rs [hunk 9]
```diff
@@ -253,6 +272,240 @@ mod send {
...
+                .unwrap()
+                .as_str()
+                .unwrap(),
+        )
+        .unwrap();
+
+        assert_eq!(expected_data, serde_json::to_value(data).unwrap());
+    }
+
+    // FIXME: Uncomment as soon as indy-utils is updated
```
Comment (berendsliedrecht) [inline libindy_vdr/tests/nym.rs:452]: indy-utils did receive an update. Can we uncomment this of do we have a to keep it commented?

File: libindy_vdr/tests/resolver.rs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        let v: serde_json::Value = serde_json::from_str(result.as_str()).unwrap();
+
+        let diddoc = &v["didDocument"];
+        let metadata = &v["didDocumentMetadata"];
+
+        assert_eq!("NYM", metadata["objectType"]);
+        assert_ne!(&serde_json::Value::Null, diddoc)
+    }
+
+    // #[rstest]
```
Comment (berendsliedrecht) [inline libindy_vdr/tests/resolver.rs:72]: Can this be removed as well?

File: libindy_vdr/tests/utils/fixtures.rs [hunk 11]
```diff
@@ -42,8 +45,42 @@ pub fn fq_my_did() -> DidValue {
 }
 
 #[fixture]
+// did:sov self-certified DID
 pub fn identity() -> Identity {
-    Identity::new(None)
+    Identity::new(None, None)
+}
+
+// FIXME: Uncomment as soon as indy-utils is updated
```
Comment (berendsliedrecht) [inline libindy_vdr/tests/utils/fixtures.rs:53]: Same comment here re `indy-utils`.

**Review by berendsliedrecht — APPROVED** [review]
Couple small items, nothing blocking.

File: libindy_vdr/src/resolver/resolver.rs (hunk 7 above)
Comment (andrewwhitehead) [reply libindy_vdr/src/resolver/resolver.rs]: I think it's short for json, since they are json nodes.

**Comment by swcurran** [comment]
w00t!!!

**Pull request merged**